### ISSUE TITLE
Snapshot repository disappears during upgrade

### ISSUE THREAD

**Issue by robin13** [elastic/elasticsearch#15529]
On my home/test cluster with one node and snapshot to local filesystem, the backup repository definition disappeared when upgrading from 2.0.0 -> 2.1.0, and again when upgrading from 2.1.0 -> 2.1.1

The snapshot is created like this:

```
curl -XPUT 'http://rcl-nas:9200/_snapshot/backup' -d '{
    "type": "fs",
    "settings":{
        "compression": true,
        "location": "/mnt/raid/elasticsearch-snapshots"
    }
}'
```

No errors referencing repositories are in the logs.

I was able to re-create the repository again with exactly the same configuration, and the previous snapshots still exist.

I have tried to reproduce this on other test cluster, but have not been able to provoke it...

**Comment by ywelsch** [comment]
I cannot reproduce this (repo is preserved).

Can you share full configs (elasticsearch.yml)?

What does your upgrade process look like?

**Comment by robin13** [comment]
I haven't been able to reproduce either, other than on the one cluster.  I have heard of others with the same experience though.
Upgrade process:
* update apt repository settings, apt-get update, apt-get upgrade
* stop elasticsearch
* uninstall and re-install plugins (watcher, license)
* start elasticsearch

Config: mostly defaults accept for memory and number of shards/replicas.

**Comment by imotov** [comment]
@robin13 did you have any templates or licenses registered in cluster? If yes, have they survived the upgrade in which repository definition disappeared? You can verify this by looking at the log file and checking if the new trail license was generated.

**Comment by robin13** [comment]
Yes - templates (logstash and beats), and license (using watcher).  License survived the upgrade.

**Comment by robin13** [comment]
Found root cause / method to reproduce:
If disk is spun down (`hdparm -Y /dev/sdX`) and Elasticsearch is restarted, the repository will be (silently - no entry in the log) removed.

**Comment by robin13** [comment]
To elaborate ^^ : The disk being spun down is the disk of the repository, not elasticsearch data.  The repository definition is removed, but the data remains (and is usable when the repository is re-created).

**Comment by ywelsch** [comment]
I think I found the issue. Seems to be the caused by the custom metadata parsing of the license plugin (still need to investigate). My observations so far:

- repositories were sometimes removed after restart of cluster
- the issue occured only when repository metadata was written after license metadata. As the order of writing custom metadata is undefined, this led to the bug occurring only sometimes (on my machine very rarely, especially NOT when I tried to reproduce it 3 days ago).

I'll have a closer look tomorrow at this.
